### ISSUE TITLE
Fabio and routing with client certificates

### ISSUE THREAD

**Issue by terusus** [eBay/fabio#174]
Hello @magiconair. Thanks a lot for Fabio! Great piece of technology!

I have the following use case.

```
internet -> fabio (https://mysite.com/...) ---> https://service-a.service.consul
                                            |
                                            --> https://service-b.service.consul
```

Now, the thing is that the services will have self-signed certificates and they will contact themselves through their Consul DNS names. And the services should rely on client certificate authentication. But Fabio has to identify itself to the incoming HTTPS request as _mysite.com_ and to the internal services as _fabio.service.consul_.

How can I properly configure Fabio to do this? And is it even worth it?

In case I drop the client certificate authentication for the user facing services how can I configure Fabio to use different certificates for internal and external traffic? Or how to properly configure the my service in order to route the encrypted traffic directly to it?

**Comment by magiconair** [comment]
fabio can handle incoming https requests in two ways:

1. terminate ssl on fabio (incl. client cert auth if needed) and forward an http request downstream
2. examine the SNI header of the TLS connection and route the encrypted tcp connection downstream based on the server name found in the SNI header. (full end-to-end encryption, tcp+sni proxy)

Fabio does not support to talk to downstream servers via https and therefore cannot authenticate via a client cert to the downstream server. I think this is what you're looking for, right?

Can you explain what you are trying to achieve?

**Comment by terusus** [comment]
> Fabio does not support to talk to downstream servers via https and therefore cannot authenticate via a client cert to the downstream server. I think this is what you're looking for, right?

That's what I was thinking of.

The goal of the entire configuration is to have 100% encrypted traffic - internal and external. Users (external traffic) authenticate with username and password. Internal traffic is authenticated with client certificates. That's why the first option does not work for me. So, option 2 it is - TCP+SNI.

Can tell more on how to make the appropriate setup with Fabio in the picture?

```
# Just this in the fabio.properties? No CS right?
proxy.addr = :443;proto=tcp+sni
```

And the the corresponding backed services must have the certificates for _mysite.com_? And register in consul with a tag `urlprefix-mysite.com/path`?

**Comment by magiconair** [comment]
Yes and `urlprefix-host/` only since the SNI extension in the TLS client hello message contains only the server name.

**Comment by magiconair** [comment]
BTW, `./fabio -proxy.addr ':443;proto=tcp+sni'` does the same job if you don't want a config file. Also note that the TCP+SNI proxy is still experimental in the sense that it works but it needs more timeouts. So if you are going to use it then let me know what your experiences are.

**Comment by terusus** [comment]
If I have more than one microservice that responds to a different portion of _mysite.com_ can I use subdomains? Like `urlprefix-api1.mysite.com/` for one service and `urlprefix-api2.mysite.com` for another service?

**Comment by magiconair** [comment]
yes

**Comment by terusus** [comment]
Thank you for the quick feedback. I'll setup my environment and I'll update you promptly regarding my experience.

**Comment by terusus** [comment]
Hello @magiconair. After quite some work it turned out that it is quite hard to securely distribute the user-facing certificates directly on the exposed backend microservices.

A much faster way to achieve the effect we needed (work with self-signed certificates internally) was to use Consul Template. Setting up a container with Consul Template and nginx took me about a day. In the template I used Consul service tag syntax that is compatible with Fabio in order to be possible to switch in the future.

I also found it quite useful to be able to expose only a specific path from a microservice. The idea is to obfuscate some parts of a service to the public.

So, if my use case is similar to the needs of other people, please, consider adding some features like:
1) different certificates for client and proxy connections
2) a tag to expose only a specific service path

**Comment by BogetC** [comment]
This feature would be so usefull.
